File: experiment-15-sonnet5/brazil/runs/language=typescript_model=sonnet-5_prompt=bdd/rep1/src/queries/stats.ts
import type { SoccerDataStore } from "../data/store.js";
import type { Competition, Match } from "../types.js";
import { filterMatches } from "./helpers.js";

export interface GoalStats {
  matchesConsidered: number;
  averageGoalsPerMatch: number;
  homeWinRate: number;
  awayWinRate: number;
  drawRate: number;
}

/** Computes goals-per-match and outcome-rate statistics (Statistical Analysis §5). */
export function calculateGoalStats(store: SoccerDataStore, options: { competition?: Competition; season?: number } = {}): GoalStats {
  const matches = filterMatches(store.matches, options).filter((m) => m.homeGoals !== null && m.awayGoals !== null);
  if (matches.length === 0) {
    return { matchesConsidered: 0, averageGoalsPerMatch: 0, homeWinRate: 0, awayWinRate: 0, drawRate: 0 };
  }

  let totalGoals = 0;
  let homeWins = 0;
  let awayWins = 0;
  let draws = 0;
  for (const match of matches) {
    const homeGoals = match.homeGoals as number;
    const awayGoals = match.awayGoals as number;
    totalGoals += homeGoals + awayGoals;
    if (homeGoals > awayGoals) homeWins += 1;
    else if (awayGoals > homeGoals) awayWins += 1;
    else draws += 1;
  }

  return {
    matchesConsidered: matches.length,
    averageGoalsPerMatch: round2(totalGoals / matches.length),
    homeWinRate: round2(homeWins / matches.length),
    awayWinRate: round2(awayWins / matches.length),
    drawRate: round2(draws / matches.length),
  };
}

/** Finds the biggest victories (largest goal difference) in the dataset, optionally scoped to a competition/season. */
export function biggestWins(store: SoccerDataStore, options: { competition?: Competition; season?: number; limit?: number } = {}): Match[] {
  const matches = filterMatches(store.matches, options).filter((m) => m.homeGoals !== null && m.awayGoals !== null);
  const sorted = [...matches].sort((a, b) => {
    const marginA = Math.abs((a.homeGoals as number) - (a.awayGoals as number));
    const marginB = Math.abs((b.homeGoals as number) - (b.awayGoals as number));
    return marginB - marginA;
  });
  return sorted.slice(0, options.limit ?? 10);
}

function round2(value: number): number {
  return Math.round(value * 100) / 100;
}

File: experiments/adrianco/experiment-2/runs/language=go_model=opus_tooling=beads/rep1/internal/query/stats.go
package query

import (
	"sort"

	"brsoccer/internal/data"
)

type GlobalStats struct {
	TotalMatches   int
	TotalGoals     int
	HomeWins       int
	AwayWins       int
	Draws          int
	AvgGoalsPerMatch float64
	HomeWinRate    float64
}

func Overall(db *data.DB, competition string, season int) GlobalStats {
	var s GlobalStats
	for _, m := range db.Matches {
		if competition != "" && !data.TeamMatches(m.Competition, competition) {
			continue
		}
		if season != 0 && m.Season != season {
			continue
		}
		s.TotalMatches++
		s.TotalGoals += m.HomeGoals + m.AwayGoals
		switch {
		case m.HomeGoals > m.AwayGoals:
			s.HomeWins++
		case m.AwayGoals > m.HomeGoals:
			s.AwayWins++
		default:
			s.Draws++
		}
	}
	if s.TotalMatches > 0 {
		s.AvgGoalsPerMatch = float64(s.TotalGoals) / float64(s.TotalMatches)
		s.HomeWinRate = float64(s.HomeWins) / float64(s.TotalMatches) * 100
	}
	return s
}

// BiggestWins returns matches sorted by goal difference.
func BiggestWins(db *data.DB, competition string, limit int) []data.Match {
	out := make([]data.Match, 0, len(db.Matches))
	for _, m := range db.Matches {
		if competition != "" && !data.TeamMatches(m.Competition, competition) {
			continue
		}
		out = append(out, m)
	}
	sort.Slice(out, func(i, j int) bool {
		di := abs(out[i].HomeGoals - out[i].AwayGoals)
		dj := abs(out[j].HomeGoals - out[j].AwayGoals)
		if di != dj {
			return di > dj
		}
		return (out[i].HomeGoals + out[i].AwayGoals) > (out[j].HomeGoals + out[j].AwayGoals)
	})
	if limit > 0 && len(out) > limit {
		out = out[:limit]
	}
	return out
}

func abs(i int) int {
	if i < 0 {
		return -i
	}
	return i
}
